### PR TITLE
Issue #27: Debugging

### PR DESCRIPTION
**Pull request by jaypha**
Add debug statements to the engine execution, and on iterator steps.

dataflow run with debugging

I specifically request comments about the extensiveness of debugging, and how to determine if debugging will occur.
OK to squash.

### CODE REVIEW

**Review by keevan — COMMENTED** [review]
I suggest changing it from debugging to mtrace as we don't need to see the stack trace that often. I think we'll be selective in where we put debugging statements, but the information provided should be a good starting point for now.

Otherwise I think it's good to merge for now and make to improvements on.

**Comment by jaypha** [comment]
I added a parameter to run.php to provide a trace level. How this gets called is not yet determined.

File: classes/execution/engine.php [hunk 1]
```diff
@@ -17,6 +17,7 @@
 namespace tool_dataflows\execution;
 
 use tool_dataflows\dataflow;
+use tool_dataflows\step\debugging;
```
Comment (keevan) [inline classes/execution/engine.php:20]: Is this required?

Comment (jaypha) [reply classes/execution/engine.php]: Ahh... No. Don't knwo why I put it in.
Removed.

File: classes/execution/engine.php [hunk 2]
```diff
@@ -206,21 +229,38 @@ public function finalise() {
...
         throw $exception;
     }
 
+    /**
+     * Display a trace message for debugging.
+     *
+     * @param string $message Message to display.
+     * @param int $tracelevel The trace level of the message.
+     */
+    public function trace(string $message, int $tracelevel = self::TRACE_ALL) {
```
Comment (brendanheywood) [inline classes/execution/engine.php:258]: Maybe this should have an optional param which takes a step object instance, and then we have a ->trace() function in the base step class which is what steps use which populates it for free and then we can do some more nuanced output depending on where we end up sending the logs to

**Comment by jaypha** [comment]
I added a trace() function to engine_step.

**Review by keevan — COMMENTED** [review]
Are you able to please change `trace` to `log` instead? It seems a bit more intuitive as it's currently serving as just a logger.

So two things:
- change `trace()` to `log()`
- change the engine property (boolean) to be `logging` instead (from trace) or similar, to make it more distinguishable from the method name.

**Review by keevan — COMMENTED** [review]
Okay, so having taken a look at this I think this needs a few small adjustments.

Here are some points of consideration:
- Logging should always be enabled. So perhaps remove the variable.
- Where things are logged will differ depending on how it is run (but it will still be logged):
  - For a dry-run, it will only log to the screen
  - For a manual run (non-dry), it will log to the screen and also to a log table linked to that instance's run.
  - A normal run triggered by webhook/cron/etc, will NOT output to the screen and will log it as above.

I think doing point 1 and enabling logging for now throughout by default will be enough for this PR. And adding a place to store the logs for the other run types mentioned above will be something added in the next PR.

**Pull request merged**